src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSTempPECompiler.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.CPS
+{
+    [Export(typeof(ITempPECompiler))]
+    internal class CPSTempPECompiler : ITempPECompiler
+    {
+        public async Task<bool> CompileAsync(IWorkspaceProjectContext context, string outputFileName, string[] filesToInclude, CancellationToken cancellationToken)
+        {
+			if (filesToInclude == null || filesToInclude.Length == 0)
+			{
+				throw new ArgumentException(nameof(filesToInclude), "Must specify some files to compile.");
+			}
+			if (string.IsNullOrWhiteSpace(outputFileName))
+			{
+				throw new ArgumentException(nameof(outputFileName), "Must specify a filename to output to.");
+			}
+
+            var snapshot = ((CPSProject)context).GetProjectSnapshot();
+            // Allow for faster checking because projects could be very large
+            var files = new HashSet<string>(filesToInclude);
+
+            // Remove all files except the ones we care about
+            var documents = snapshot.Documents;
+            foreach (var document in documents)
+            {
+                if (!files.Contains(document.FilePath))
+                {
+                    snapshot = snapshot.RemoveDocument(document.Id);
+                }
+            }
+
+            // We want to produce a DLL regardless of project type
+            snapshot = snapshot.WithCompilationOptions(snapshot.CompilationOptions.WithOutputKind(OutputKind.DynamicallyLinkedLibrary));
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var compilation = await snapshot.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            if (compilation.GetDiagnostics(cancellationToken).HasAnyErrors())
```

Comment:
Do you care about warnings as errors interfering with this?

Reply:
This is a good point. I can add `.WithGeneralDiagnosticOption(ReportDiagnostic.Suppress).WithSpecificDiagnosticOptions(null)` but I'm wondering if chasing options to unset just in case is a bad idea. What if the current project has strong naming, will that mean the TempPE DLL does too? 

Is there a way to get a "default" CompilationOptions in a language agnostic way? We only really want references and defines.

Reply:
> What if the current project has strong naming, will that mean the TempPE DLL does too?

What did the old thing do?

You can do `project.LanguageServices.GetRequiredService<ICompilationFactoryService>().GetDefaultOptions()` if you want to get the true default, but yes, then that means you're chasing this in the other way.

> We only really want references and defines.

What all gets passed in as the file list, in practice? Are they ever user files (or at least contain user files), or just truely autogenerated stuff? Because /langversion, /unsafe, and a few other things also fundamentally control whether you can interpret user code or not.

Reply:
Strictly speaking there could be user code because of customizations to the project file, but normally I would expect generated files only.
The old TempPE compiler just passes in `/target:library /debug- /nostdlib+ /d:.. /d:.. /r:.. /r:..`.

I think starting from default is probably a better fit for this, thanks for the code snippet.